src/Avalonia.Controls/DateTimePickers/DatePicker.cs

```diff
@@ -0,0 +1,517 @@
+            {
+                _container.ColumnDefinitions.Add(new ColumnDefinition(_monthIndex < _dayIndex ? 132 : 78, GridUnitType.Star));
+                _container.ColumnDefinitions.Add(new ColumnDefinition(0, GridUnitType.Auto));
+                _container.ColumnDefinitions.Add(new ColumnDefinition(_monthIndex < _dayIndex ? 78 : 132, GridUnitType.Star));
+
+                _monthText.IsVisible = true;
+                _dayText.IsVisible = true;
+                _yearText.IsVisible = false;
+                _spacer1.IsVisible = true;
+                _spacer2.IsVisible = false;
+
+                Grid.SetColumn(_monthText, _monthIndex < _dayIndex ? 0 : 2);
+                Grid.SetColumn(_dayText, _monthIndex < _dayIndex ? 2 : 0);
+                Grid.SetColumn(_spacer1, 1);
+            }
+            else if (showMonth && !showDay && showYear) //Month and Year Only
+            {
+                _container.ColumnDefinitions.Add(new ColumnDefinition(_monthIndex < _yearIndex ? 132 : 78, GridUnitType.Star));
+                _container.ColumnDefinitions.Add(new ColumnDefinition(0, GridUnitType.Auto));
+                _container.ColumnDefinitions.Add(new ColumnDefinition(_monthIndex < _yearIndex ? 78 : 132, GridUnitType.Star));
+
+                _monthText.IsVisible = true;
+                _dayText.IsVisible = false;
+                _yearText.IsVisible = true;
+                _spacer1.IsVisible = true;
+                _spacer2.IsVisible = false;
+
+                Grid.SetColumn(_monthText, _monthIndex < _yearIndex ? 0 : 2);
+                Grid.SetColumn(_yearText, _monthIndex < _yearIndex ? 2 : 0);
+                Grid.SetColumn(_spacer1, 1);
+            }
+            else if (!showMonth && showDay && showYear) //Day and Year Only
+            {
+                _container.ColumnDefinitions.Add(new ColumnDefinition(78, GridUnitType.Star));
+                _container.ColumnDefinitions.Add(new ColumnDefinition(0, GridUnitType.Auto));
+                _container.ColumnDefinitions.Add(new ColumnDefinition(78, GridUnitType.Star));
+
+                _monthText.IsVisible = false;
+                _dayText.IsVisible = true;
+                _yearText.IsVisible = true;
+                _spacer1.IsVisible = true;
+                _spacer2.IsVisible = false;
+
+                Grid.SetColumn(_yearText, _dayIndex < _yearIndex ? 2 : 0);
+                Grid.SetColumn(_dayText, _dayIndex < _yearIndex ? 0 : 2);
+                Grid.SetColumn(_spacer1, 1);
+            }
+            else if (showMonth && showDay && showYear) //All Visible
+            {
+                bool isMonthFirst = _monthIndex < _dayIndex && _monthIndex < _yearIndex;
+                bool isMonthSecond = (_monthIndex > _dayIndex && _monthIndex < _yearIndex) ||
+                    (_monthIndex < _dayIndex && _monthIndex > _yearIndex);
+
+                _container.ColumnDefinitions.Add(new ColumnDefinition(isMonthFirst ? 138 : 78, GridUnitType.Star));
```

Comment:
Seems like we use these constants all-around? can we sort these stuff in the whole PR and name these magical constants?

Reply:
IIRC, these are used only here and similar logic is reused in `DatePickerPresenter`. I wrote the two classes to be independent, similar to how all this works in WinUI (you can use a DatePickerFlyout without a DatePicker).